### PR TITLE
[LLM] Example for Serving Gemma

### PR DESCRIPTION
**Pull request by Michaelvll**
TODO (future):
- [ ] add finetuning as well

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch -c gemma llm/gemma/serve.yaml --cloud gcp --env HF_TOKEN="xxx"`
  - [x] `curl` the model with command in readme
    ```bash
    IP=$(sky status --ip gemma)
    curl -L http://$IP:8000/v1/completions   -H "Content-Type: application/json"   -d '{
          "model": "google/gemma-7b",
          "prompt": "My favourite condiment is",
          "max_tokens": 25
      }'
    ```
  - [x] `sky serve up -n gemma serve.yaml --env HF_TOKEN`
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: llm/gemma/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+## Serve Gemma on any Cloud
+
+Serving Gemma on any cloud is easy with SkyPilot. We can find the [serve.yaml](serve.yaml) in this directory, which can be used to start hosting the model on any cloud with a single command.
+
+### Prerequsite
+
+1. Apply for the access to the Gemma model
+
+Go to the [application page](https://huggingface.co/google/gemma-7b) and apply for the access to the model weights.
```
Comment (concretevitamin) [inline llm/gemma/README.md:14]:
```suggestion
Go to the [application page](https://huggingface.co/google/gemma-7b) and click **Acknowledge license** to apply for access to the model weights.
```

File: llm/gemma/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+2. Get the access token from huggingface
+
+Generate a read-only access token on huggingface [here](https://huggingface.co/settings/token), and make sure your huggingface account can access the Gemma models [here](https://huggingface.co/google/gemma-7b).
+
+
+### Host on a Single Instance
+
+We can host the model with a single instance:
+```bash
+sky launch -c gemma serve.yaml --env HF_TOKEN="xxx"
```
Comment (concretevitamin) [inline llm/gemma/README.md:26]: nit: How about `HF_TOKEN=xxx sky launch ...`?

File: llm/gemma/serve.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  pip install vllm==0.3.2
+  pip install transformers==4.38.0
+  python -c "import huggingface_hub; huggingface_hub.login('${HF_TOKEN}')"
+
+run: |
+  conda activate gemma
+  export PATH=$PATH:/sbin
+  # --max-model-len is set to 1024 to avoid taking too much GPU memory on L4 and
+  # A10g with small memory.
+  python -u -m vllm.entrypoints.openai.api_server \
```
Comment (concretevitamin) [inline llm/gemma/serve.yaml:43]: nit: 2 space indent?

File: llm/gemma/serve.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    path: /v1/chat/completions
+    post_data:
+      model: $MODEL_NAME
+      messages:
+        - role: user
+          content: Hello! What is your name?
+      max_tokens: 1
+    initial_delay_seconds: 1200
+  replica_policy:
+    min_replicas: 2
```
Comment (concretevitamin) [inline llm/gemma/serve.yaml:17]: nit: Shall we use the simpler `replicas: 2` syntax?

File: README.md [hunk 5]
```diff
@@ -27,6 +27,7 @@
 
 ----
 :fire: *News* :fire:
+- [Feb, 2024] Deploying and scaling [**Gemma**](https://blog.google/technology/developers/gemma-open-models/) wiht SkyServe: [**example**](./llm/gemma/)
```
Comment (concretevitamin) [inline README.md:30]:
```suggestion
- [Feb, 2024] Deploying and scaling [**Gemma**](https://blog.google/technology/developers/gemma-open-models/) with SkyServe: [**example**](./llm/gemma/)
```

File: llm/gemma/README.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      "model": "google/gemma-7b",
+      "prompt": "My favourite condiment is",
+      "max_tokens": 25
+  }' | jq .
+```
+
+### Scale the Serving with SkyServe
+
+
+We can scale the model serving across multiple instances, regions and clouds with SkyServe:
```
Comment (concretevitamin) [inline llm/gemma/README.md:45]:
```suggestion
Using the same YAML, we can easily scale the model serving across multiple instances, regions and clouds with SkyServe:
```

File: llm/gemma/README.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+```bash
+sky serve up -n gemma serve.yaml --env HF_TOKEN="xxx"
+```
+
```
Comment (concretevitamin) [inline llm/gemma/README.md:49]: Maybe add something like
```
> Notice the only change is from `sky launch` to `sky serve up`. The same YAML can be used without changes.
```

**Review by concretevitamin — COMMENTED** [review]
Awesome, @Michaelvll! Sending some comments first while it's running in background.

File: llm/gemma/README.md (hunk 2 above)
Comment (concretevitamin) [reply llm/gemma/README.md:26]: (same for sky serve up below?)

File: llm/gemma/serve.yaml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+envs:
+  MODEL_NAME: google/gemma-7b
+  HF_TOKEN: <your-huggingface-token> # TODO: Replace with huggingface token
+
+resources:
```
Comment (concretevitamin) [inline llm/gemma/serve.yaml:23]: Azure is taking a long time without success (quota failover), so I changed this to
```
resources:
  accelerators: {L4, A10g, A10, L40, A40, A100, A100-80GB}
  ports: 8000
  disk_tier: best
  any_of:
    - cloud: aws
    - cloud: gcp
```
and got
```

  File "/Users/zongheng/Dropbox/workspace/riselab/sky-computing/sky/serve/core.py", line 88, in up
    raise ValueError(f'Got multiple clouds: {requested_cloud} and '
ValueError: Got multiple clouds: GCP and AWS in different resources. Please specify single cloud instead.
```
for `sky serve up`. Is this expected?

File: llm/gemma/README.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+Google released [Gemma](https://blog.google/technology/developers/gemma-open-models/) and has made a big wave in the AI community.
+It opens the opportunity for the open-source community to serve and finetune private Gemini.
+
+## Serve Gemma on any Cloud
+
+Serving Gemma on any cloud is easy with SkyPilot. We can find the [serve.yaml](serve.yaml) in this directory, which can be used to start hosting the model on any cloud with a single command.
+
+### Prerequsite
+
```
Comment (concretevitamin) [inline llm/gemma/README.md:11]: Shall we add a pointer to install nightly docs?

File: llm/gemma/serve.yaml [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  conda activate gemma
+  export PATH=$PATH:/sbin
+  # --max-model-len is set to 1024 to avoid taking too much GPU memory on L4 and
+  # A10g with small memory.
+  python -u -m vllm.entrypoints.openai.api_server \
+                --host 0.0.0.0 \
+                --model $MODEL_NAME \
+                --tensor-parallel-size $SKYPILOT_NUM_GPUS_PER_NODE \
+                --max-model-len 1024 | tee ~/openai_api_server.log
+
```
Comment (concretevitamin) [inline llm/gemma/serve.yaml:48]: Maybe add a short section on changing `--model` to instruction tuned versions? Which have been added to various vendors.

File: llm/gemma/README.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
+# Serve Your Gemma on Any Cloud
+
+Google released [Gemma](https://blog.google/technology/developers/gemma-open-models/) and has made a big wave in the AI community.
+It opens the opportunity for the open-source community to serve and finetune private Gemini.
+
+## Serve Gemma on any Cloud
+
+Serving Gemma on any cloud is easy with SkyPilot. We can find the [serve.yaml](serve.yaml) in this directory, which can be used to start hosting the model on any cloud with a single command.
```
Comment (concretevitamin) [inline llm/gemma/README.md:8]:
```suggestion
Serving Gemma on any cloud is easy with SkyPilot. With [serve.yaml](serve.yaml) in this directory, you host the model on any cloud with a single command.
```

File: llm/gemma/README.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+### Prerequsite
```
Comment (concretevitamin) [inline llm/gemma/README.md:10]:
```suggestion
### Prerequisites
```

File: llm/gemma/README.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+1. Apply for the access to the Gemma model
```
Comment (concretevitamin) [inline llm/gemma/README.md:12]:
```suggestion
1. Apply for access to the Gemma model
```

**Review by concretevitamin — APPROVED** [review]
Works like a charm on by `launch` and `serve up`. LGTM.

File: llm/gemma/README.md (hunk 2 above)
Comment (Michaelvll) [reply llm/gemma/README.md:26]: Specifying the `HF_TOKEN` as a env directly won't work in our `sky launch` : )

Comment (concretevitamin) [reply llm/gemma/README.md:26]: I meant `HF_TOKEN=xxx sky launch --env HF_TOKEN ...`.

File: llm/gemma/serve.yaml (hunk 8 above)
Comment (Michaelvll) [reply llm/gemma/serve.yaml:22]: This seems to be an issue that needs to be fixed @cblmemo @MaoZiming ?

File: llm/gemma/serve.yaml (hunk 10 above)
Comment (Michaelvll) [reply llm/gemma/serve.yaml:47]: Good point! Changed it to `gemma-7b-it` : )

File: llm/gemma/serve.yaml (hunk 8 above)
Comment (cblmemo) [reply llm/gemma/serve.yaml:22]: Good catch! Fixed in #3226. Thanks!

**Pull request merged**